plugins/redfish/fu-redfish-backend.c

```diff
@@ -0,0 +1,391 @@
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include <string.h>
+
+#include <fwupdplugin.h>
+
+#include "fu-redfish-backend.h"
+#include "fu-redfish-common.h"
+#include "fu-redfish-device.h"
+#include "fu-redfish-request.h"
+#include "fu-redfish-smbios.h"
+
+struct _FuRedfishBackend
+{
+	FuBackend		 parent_instance;
+	gchar			*hostname;
+	gchar			*username;
+	gchar			*password;
+	guint			 port;
+	gchar			*update_uri_path;
+	gchar			*push_uri_path;
+	gboolean		 use_https;
+	gboolean		 cacheck;
+	gint64			 max_image_size;	/* bytes */
+	FuRedfishBackendKind	 kind;
+	GHashTable		*request_cache;		/* str:GByteArray */
+};
+
+G_DEFINE_TYPE (FuRedfishBackend, fu_redfish_backend, FU_TYPE_BACKEND)
+
+const gchar *
+fu_redfish_backend_kind_to_string (FuRedfishBackendKind kind)
+{
+	if (kind == FU_REDFISH_BACKEND_UPDATE_KIND_LEGACY)
+		return "mooo";
+	if (kind == FU_REDFISH_BACKEND_UPDATE_KIND_MULTIPART)
+		return "multipart";
+//	if (kind == FU_REDFISH_BACKEND_UPDATE_KIND_OEM_LENOVO)
+//		return "oem-lenovo";
+	return NULL;
+}
+
+FuRedfishRequest *
+fu_redfish_backend_request_new (FuRedfishBackend *self)
+ {
+	FuRedfishRequest *request = g_object_new (FU_TYPE_REDFISH_REQUEST, NULL);
+	CURL *curl;
+	CURLU *uri;
+	g_autofree gchar *user_agent = NULL;
+	g_autofree gchar *port = g_strdup_printf ("%u", self->port);
+
+	/* set the cache location */
+	fu_redfish_request_set_cache (request, self->request_cache);
+
+	/* set up defaults */
+	curl = fu_redfish_request_get_curl (request);
+	uri = fu_redfish_request_get_uri (request);
+	curl_url_set (uri, CURLUPART_SCHEME, self->use_https ? "https" : "http", 0);
+	curl_url_set (uri, CURLUPART_HOST, self->hostname, 0);
+	curl_url_set (uri, CURLUPART_PORT, port, 0);
+	curl_easy_setopt (curl, CURLOPT_CURLU, uri);
+
+	/* since DSP0266 makes Basic Authorization a requirement,
+	* it is safe to use Basic Auth for all implementations */
+	curl_easy_setopt (curl, CURLOPT_HTTPAUTH, (glong) CURLAUTH_BASIC);
+	curl_easy_setopt (curl, CURLOPT_USERNAME, self->username);
+	curl_easy_setopt (curl, CURLOPT_PASSWORD, self->password);
+
+	/* setup networking */
+	user_agent = g_strdup_printf ("%s/%s", PACKAGE_NAME, PACKAGE_VERSION);
+	curl_easy_setopt (curl, CURLOPT_USERAGENT , user_agent);
+	curl_easy_setopt (curl, CURLOPT_CONNECTTIMEOUT, 60L);
+	if (!self->cacheck)
+		curl_easy_setopt (curl, CURLOPT_SSL_VERIFYPEER, 0L);
+
+	/* success */
+	return request;
+ }
+
+static gboolean
+fu_redfish_backend_coldplug_member (FuRedfishBackend *self,
+				    JsonObject *member,
+				    GError **error)
+{
+	g_autoptr(FuDevice) dev = fu_redfish_device_new_from_object (self, member);
+	g_autoptr(FuDeviceLocker) locker = NULL;
+
+	locker = fu_device_locker_new (dev, error);
+	if (locker == NULL)
+		return FALSE;
+	if (self->max_image_size != 0)
+		fu_device_set_firmware_size_max (dev, (guint64) self->max_image_size);
+	//fu_device_add_private_flag (dev, FU_REDFISH_DEVICE_FLAG_USE_MULTIPART);
```

Comment:
how do you know when this is needed?

Reply:
My heuristic so far is USE_MULTIPART if MultipartHttpPushUri is defined, or use legacy if not. I'm not sure if that's good enough or if we'll have to quirk that in the conf file.

Reply:
Removed.